libdevcore/db.h

```diff
@@ -21,16 +21,80 @@
 
 #pragma once
 
+#include "Exceptions.h"
 #include "dbfwd.h"
 
-#pragma warning(push)
-#pragma warning(disable: 4100 4267)
-#if ETH_ROCKSDB
-#include <rocksdb/db.h>
-#include <rocksdb/write_batch.h>
-#else
-#include <leveldb/db.h>
-#include <leveldb/write_batch.h>
-#endif
-#pragma warning(pop)
-#define DEV_LDB 1
+#include <memory>
+#include <string>
+
+namespace dev
+{
+namespace db
+{
+
+// Transaction implements database transaction for a specific concrete database
+// implementation. The transaction implementation must call rollback in its
+// destructor if commit has not yet been called (making sure to catch any
+// exceptions).
+class Transaction
+{
+public:
+	virtual ~Transaction() = default;
+
+	virtual void insert(Slice const& _key, Slice const& _value) = 0;
+	virtual void kill(Slice const& _key) = 0;
+
+	virtual void commit() = 0;
+	virtual void rollback() = 0;
+
+protected:
+	Transaction() = default;
+	// Noncopyable
+	Transaction(const Transaction&) = delete;
+	Transaction& operator=(const Transaction&) = delete;
+	// Movable
+	Transaction(Transaction&&) = default;
+	Transaction& operator=(Transaction&&) = default;
+};
+
+class DB
+{
+public:
+	virtual ~DB() = default;
+	virtual std::string lookup(Slice const& _key) const = 0;
+	virtual bool exists(Slice const& _key) const = 0;
+	virtual void insert(Slice const& _key, Slice const& _value) = 0;
+	virtual void kill(Slice const& _key) = 0;
+	virtual std::unique_ptr<Transaction> begin() = 0;
+
+	// A database must implement the `forEach` method that allows the caller
+	// to pass in a function `f`, which will be called with the key and value
+	// of each record in the database. If `f` returns false, the `forEach`
+	// method must return immediately.
+	virtual void forEach(std::function<bool(Slice const&, Slice const&)> f) const = 0;
+
+	virtual void print(std::ostream& out) const
+	{
+		forEach([&out](Slice const& key, Slice const& value) {
+			out << toHex(key) << " => " << toHex(value) << '\n';
+			return static_cast<bool>(out);
+		});
+	}
+};
+
+inline std::ostream& operator<<(std::ostream& out, const DB& database)
+{
+	database.print(out);
+	return out;
+}
+
+struct FailedToOpenDB: virtual Exception { using Exception::Exception; };
```

Comment:
I like you didn't use Boost here. I believe using Boost Exceptions does not give you a lot, but obscures code, especially `BOOST_THROW_EXCEPTION`.

But, I'm not sure it is good to have so much exception types. Can we try to have single exception type for database this time?

@gumb0 ?

Reply:
What do you mean they didn't use boost, `Exception` is derived from `boost::exception`?

In general I think having a class for each kind of exception was not a very good idea, but I'm not sure about changing this practice now.
Different errors still should be distinguishable, we'd need to create some `enum` for error code if we use one exception class for db errors...

Reply:
`BOOST_THROW_EXCEPTION` gives us output of source code location of throw in debug build, this is valuable

Reply:
Also not using `DEV_SIMPLE_EXCEPTION` macro to define exceptions here goes against the practice of surrounding code and this way name of exception class won't be output

Reply:
Actually I took the errors here from the since-removed WhisperDB files.